keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -0,0 +1,705 @@
+
+#### Diagnostics
+
+Clusters within my ClusterSet send logs/metrics to a common monitoring solution and I need to be able to identify the cluster from which a given set of events originated.
+
+
+#### Joining or moving between ClusterSets
+
+I want the ability to add a previously-isolated cluster to a ClusterSet, or to move a cluster from one ClusterSet to another and be aware of this change.
+
+### `ClusterClaim` CRD
+
+The `ClusterClaim` resource provides a way to store identification related, cluster scoped information for multi-cluster tools while creating flexibility for implementations. A cluster may have multiple `ClusterClaim`s, each holding a different identification related value. Each claim contains the following information:
+
+*   **Name** - a well known or custom name to identify the claim.
+*   **Value** - a claim-dependent string.
+
+
+### Well known claims
+
+The requirements below use the keywords **must, should,** and **may** purposefully in accordance with [RFC-2119](https://tools.ietf.org/html/rfc2119).
+
+
+#### Claim: `id.k8s.io`
+
+Contains a unique identifier for the containing cluster.
+
+
+##### Uniqueness
+
+*   The identifier **must** be unique within the ClusterSet to which its cluster belongs for the duration of the cluster’s membership.
+*   The identifier **should** be unique beyond the ClusterSet within the scope of expected use.
+*   The identifier **may** be globally unique beyond the scope of its ClusterSet.
+*   The identifier **may** be unique beyond the span of its cluster’s membership and lifetime.
+
+
+##### Lifespan
+
+*   The identifier **must** exist and be immutable for the duration of a cluster’s membership in a ClusterSet, and as long as a `clusterset.k8s.io` claim exists.
+*   The identifier **must** exist for the lifespan of a cluster.
+*   The identifier **may** be immutable for the lifespan of a cluster.
+
+
+##### Contents
+
+*   The identifier **must** be a valid RFC-1123 DNS label.
+    *   Following the most restrictive standard naming constraint ensures maximum usefulness and portability.
+    *   Can be used as a component in MCS DNS.
+*   The identifier **may** be a human readable description of its cluster.
+
+
+##### Consumers
+
+*   Can rely on the identifier existing, unmodified for the entire duration of its membership in a ClusterSet.
+*   **Should** watch the `id.k8s.io` claim to handle potential changes if they live beyond the ClusterSet membership.
+*   **May** rely on the existence of an identifier for clusters that do not belong to a ClusterSet so long as the implementation provides one.
+
+
+#### Claim: `clusterset.k8s.io`
+
+Contains an identifier that relates the containing cluster to the ClusterSet in which it belongs.
+
+
+##### Lifespan
+
+*   The identifier **must** exist and be immutable for the duration of a cluster’s membership in a ClusterSet.
```

Comment:
given that clusterset kind of already exists (people are doing MC stuff, aanyway) I do not think this can practically be *must* (at least wrt existence)

Reply:
Same.

Reply:
Perhaps the qualifier is "to be considered a ClusterClaim implementation".

We can talk about migration to ClusterClaims in another section, but if we want to allow e.g. an MCS implementation to build on ClusterClaims it seems like there needs to be some safe assumptions about lifetime. That could be something like: when adopting claims, a cluster set is not considered ready until each cluster has the required claims.

Reply:
Is it sufficient to say that some features depend on some "facts" to be present and if they are not present, those features may not work?

Reply:
Added explicit language about this in ede2bea50f290e1ae09e00df5a0876b16518c820